exercises/sgf-parsing/description.md

```diff
@@ -53,6 +53,18 @@ A key can have multiple values associated with it. For example:
 
 Here `AB` (add black) is used to add three black stones to the board.
 
+According the the SGF specification `\` is the escape character. When `\` is
+encountered during parsing, it indicates that the following character should be
+printed as is. The exception to this is whitespace, with newline, carriage
+return and tab encoded as `"\n"`, `"\r"` and `"\t"` respectively. When parsing
+you can expect some characters to _always_ be escaped, including parentheses,
+square brackets, semicolons and colons.
+
+When parsing SGF files whitespace gets special handling, with the exception of
+newlines any whitespace character encountered is replaced with a single space
+character irrespective of if the original character is found in its escaped
+from or its unescaped form.
```

Comment:
While I think the idea of this distinction between "its escaped form or its unescaped form" contradicts https://www.red-bean.com/sgf/sgf4.html, fixing this would be a larger undertaking and also involves changing the case currently in the canonical-data.json, so I'll be silent about it for now and refrain from making that a part of this PR.

Reply:
I thought it was consistent with the spec.

The spec is not explicit but it states
- "White spaces other than linebreaks are converted to space"
- "Any char following "\" is inserted verbatim (exception: whitespaces still have to be converted to space!)."
- "Following chars have to be escaped, when used in Text: "]", "\" and ":" (only if used in compose data type)."

which seems to indicate:
1. that escaped WS is interpreted as WS (so`"\t"` is a tab and not an escaped `t`) and should be replaced with a space character.
2. that whitespace does not have to be escaped so unescaped whitespace would also need to be converted to spaces (save for line break)

Reply:
You have already pointed out the crucial statement: "Any char following `"\"` is inserted verbatim". Therefore, the SGF parser, when seeing the sequence of characters `\t`, should not emit a tab or a space but just a `t`.

Since we have a test in the canonical-data.json that does otherwise, **our canonical-data.json is wrong**. But, we can fix that later - since the test has been wrong for years, no additional harm is caused by it remaining wrong for some time longer.

The improvement to the canonical-data.json would probably be having individual tests for some of the individual required behaviours (replace whitespace with a space, replace an escaped newline with nothing at all, all other escaped characters).

So, let's not worry about it for this PR and just stick with what's there.

Reply:
`"Any char following "\" is inserted verbatim (exception: whitespaces still have to be converted to space!)."` you need to read what they have in the parentheses, if `"\t"` was to be replaced with `"t"` then how do you understand that "exception"?

Reply:
A backslash followed by a tab should become a space. And a backslash followed by a t should become a t.

Reply:
Important to distinguish between:

* strings as they appear in a in JSON or a given programming language's string literal
* strings as they are presented to the SGF parser

If the string presented to the SGF parser contains a backslash followed by a tab (which will typically be written in JSON and/or in a language's string literal as `"\\\t"` that should be converted to a space, because "(exception: whitespaces still have to be converted to space)".

If the string presented to the SGF parser contains a backslash followed by a t (which will typically be written in JSON and/or in a language's string literal as "\\t" that should be converted to a t, because "Any char following `"\"` is inserted verbatim"

Reply:
Important to distinguish between:

* strings as they appear in a in JSON or a given programming language's string literal
* strings as they are presented to the SGF parser

If the string presented to the SGF parser contains a backslash followed by a tab (which will typically be written in JSON and/or in a language's string literal as `"\\\t"` that should be converted to a space, because "(exception: whitespaces still have to be converted to space)".

If the string presented to the SGF parser contains a backslash followed by a t (which will typically be written in JSON and/or in a language's string literal as `"\\t"` that should be converted to a t, because "Any char following `"\"` is inserted verbatim"

Reply:
If the string presented to the SGF parser contains a tab (written as "\t"), that should be converted to a space because "White spaces other than linebreaks are converted to space"